File: java-code/src/main/java/com/vd/study/java_code/observer/Observable.java
package com.vd.study.java_code.observer;

import java.util.HashSet;
import java.util.Set;

public class Observable<T> {

    private T value;

    private final Set<Observer<T>> observers = new HashSet<>();

    public Observable(T value) {
        this.value = value;
    }

    public void setValue(T value) {
        this.value = value;
        notifyObservers();
    }

    public T getValue() {
        return value;
    }

    public void register(Observer<T> observer) {
        observers.add(observer);
    }

    public void unregister(Observer<T> observer) {
        observers.remove(observer);
    }

    private void notifyObservers() {
        for (Observer<T> observer : observers) {
            observer.onChanged(value);
        }
    }
}

File: kotlin-code/src/main/java/com/vd/study/kotlin_code/observer/Observable.kt
package com.vd.study.kotlin_code.observer

open class Observable<T>(value: T) {

    var value: T = value
        set(value) {
            field = value
            notifyObservers()
        }

    private val observers = mutableSetOf<Observer<T>>()

    open fun register(observer: Observer<T>) {
        observers += observer
    }

    open fun unregister(observer: Observer<T>) {
        observers -= observer
    }

    private fun notifyObservers() {
        observers.forEach {
            it.onChanged(value)
        }
    }
}

fun <T> Observable<T>.observe(
    observer: (value: T) -> Unit
): Observer<T> {
    val observerObject = object : Observer<T> {
        override fun onChanged(value: T) {
            observer(value)
        }
    }
    this.register(observerObject)
    return observerObject
}